packages/sync-server/src/sync/snapshotOutgoingChanges.ts

```diff
@@ -0,0 +1,96 @@
+import log from 'winston';
+
+import { getSnapshotTableName, NON_SYNCING_TABLES, SYNC_SESSION_DIRECTION } from '@tupaia/sync';
+import { ModelRegistry, TupaiaDatabase } from '@tupaia/database';
+import { SyncServerConfig } from '../types';
+
+type SnapshotOutgoingChangesResult = {
+  maxId: string;
+  count: string;
+};
+
+export const snapshotOutgoingChanges = async (
+  database: TupaiaDatabase,
+  models: ModelRegistry,
+  since: number,
+  sessionId: string,
+  deviceId: string,
+  projectIds: string[],
+  config: SyncServerConfig,
+): Promise<number> => {
+  let fromId = '';
+  let totalCount = 0;
+  const snapshotTableName = getSnapshotTableName(sessionId);
+  const CHUNK_SIZE = config.maxRecordsPerSnapshotChunk;
+  const avoidRepull = config.lookupTable.avoidRepull;
+  const recordTypes = Object.values(models)
+    .filter(m => !NON_SYNCING_TABLES.includes(m.databaseRecord) && !!m.databaseRecord)
+    .map(m => m.databaseRecord);
+
+  while (fromId != null) {
+    const [{ maxId, count }] = (await database.executeSql(
+      `
+      WITH inserted AS (
+        INSERT INTO ${snapshotTableName} (
+          sync_lookup_id,
+          direction,
+          record_type,
+          record_id,
+          saved_at_sync_tick,
+          data
+        )
+        SELECT
+          id,
+          '${SYNC_SESSION_DIRECTION.OUTGOING}',
+          record_type,
+          record_id,
+          updated_at_sync_tick,
+          data
+        FROM
+          sync_lookup
+        WHERE updated_at_sync_tick > ?
+        ${fromId ? `AND id > ?` : ''}
+        AND (
+          project_ids IS NULL
+          OR
+          project_ids::text[] @> ARRAY[${projectIds.map(p => `?`).join(',')}]
+        )
+        AND record_type IN (${recordTypes.map(r => `?`).join(',')})
+        ${
+          avoidRepull && deviceId
+            ? 'AND (pushed_by_device_id <> ? OR pushed_by_device_id IS NULL)'
+            : ''
+        }
+        ORDER BY id
+        LIMIT ?
+        RETURNING sync_lookup_id
+      )
+      SELECT MAX(sync_lookup_id) as "maxId",
+        count(*) as "count"
```

Comment:
Seeing as it’s so consistently formatted in the rest of this template:

```suggestion
      SELECT
        MAX(sync_lookup_id) as "maxId",
        COUNT(*) as "count"
```